### ISSUE TITLE
perf(cuda): witness --cuda-graph decode replay at 27B on A100; auto-enable by model size if it wins (#483)

### ISSUE THREAD

**Issue by anthony-chaudhary** [anthony-chaudhary/fak#932]
## What

Witness the CUDA-graph decode-replay path (#483) at **27B on an A100**, and — only if it
wins there — make the gate **model-size-aware** so it auto-enables for large single-stream
decode instead of staying behind a flag.

## Why now

The path is fully implemented and wired through the HAL (`internal/model/hal.go`
GraphBegin/GraphEndLaunch; `internal/compute/cuda_kernels.cu` keeps `g_exec` and patches it
with `cudaGraphExecUpdate`). It is gated OFF with the note *"per-token capture is a measured
no-win"* — but that measurement was on a **tiny 0.5B model on an L4**, where per-kernel launch
overhead is already a small fraction of token time. A **27B single-stream decode on an A100**
is a different regime: far more, larger kernels per token, and the decode is memory-bandwidth
bound, so launch gaps that leave the memory system idle cost real tok/s. The lever is now one
flag away (`fak serve --cuda-graph`, bf89b5a) — it has never been measured at the scale where
it should matter.

## Steps / acceptance

- [ ] On an A100 serving Qwen3.6-27B q4_k (`tools/qwen36_a100_fak_serve.sh`), capture warm
      steady-state decode tok/s with `CUDA_GRAPH=0` vs `CUDA_GRAPH=1` (discard the cold token).
      Record the before/after as a witness under `experiments/gpu/`.
- [ ] If graph-on wins by a meaningful margin, replace the hardcoded `graphEnabled` default
      with a **size-aware** gate (e.g. enable when resident params / per-token launch count
      exceed a threshold), keeping small models OFF (the witnessed 0.5B no-win) — and keep the
      explicit `--cuda-graph` / `FAK_CUDA_GRAPH` override. Add a GPU-free unit test for the
      threshold decision (the seam in `internal/compute/graph_cuda.go`).
- [ ] If it does **not** win at 27B/A100 either, record that as the witness and document the
      flag as a no-op-on-this-hardware lever (honest negative result), closing the loop.

Do **not** flip the default on a guess — this issue exists to ground the flip (or the
non-flip) in a measured number, per the repo's witnessed-claim discipline.

Parent: the Qwen3.6-27B-on-A100 epic. Mechanism: #483. Sibling lever (Metal): #67.

**Comment by anthony-chaudhary** [comment]
**Witnessed on a live A100-40GB (2026-06-27), Qwen2.5-Coder-14B Q8-resident via `fak serve --backend cuda`:**

- **Baseline (no graph): ~6.67 tok/s** (56 completion tokens / 8.39 s; ~6% of the A100 BW ceiling — heavily launch/op bound, exactly the regime #483 targets, unlike the 0.5B/L4 no-win).
- **`--cuda-graph` (the flag from bf89b5a): the decode turn CRASHES**, not speeds up:
  ```
  fak-cuda: cudaMalloc(4194304) failed: operation not permitted when stream is capturing
  oom-retry site=kv-key-prealloc layer 0 → in-kernel GPU OOM (kv_cache allocation failed)
  ```

**Root cause:** the KV cache is allocated **lazily during the first captured decode token** (`kv-key-prealloc`, layer 0), but the graph path's documented precondition is *no cudaMalloc during capture* (cuda.go GraphBegin). The lazy KV alloc violates it → the alloc fails mid-capture → the turn OOMs.

**Fix (the real #483/#932 work):** pre-allocate the FULL KV cache to `--context-budget-tokens` BEFORE the first graph capture (and warm the buffer pool), so the captured decode does zero cudaMalloc. Until then `--cuda-graph` must stay OFF for the in-kernel serve (default is OFF, so the shipped default is safe). This is the concrete precondition gap to close before the graph lever can be witnessed for tok/s.

**Comment by anthony-chaudhary** [comment]
**Empirical refinement (live A100, 14B):** the fix is NOT only sizing `cudaKVMaxPos` (=1024) to the context budget. Tested `--cuda-graph` with `--context-budget-tokens 512` (well UNDER the 1024 cap): it STILL crashes at `site=kv-key-prealloc layer 0` — i.e. `NewKV`'s prealloc `dallocKV` (cuda.go:1172-1176) itself runs DURING capture, even on the warmup turn. So #932 has TWO parts: (1) `NewKV` must complete before the first `GraphBegin` (ordering), and (2) the fixed cap must be ≥ the session's max positions (so `growAppend` never reallocs mid-capture). The perf-model agent's '90% launch-bound, graph = 8-15x' makes this the top lever — see #483.

**Comment by anthony-chaudhary** [comment]
**ROOT CAUSE — found by rebuilding with cudaKVMaxPos=8192 and reading the full crash sequence on the live A100 (14B, warmup turn, fresh process where non-graph works fine):**

The cap-size fix is necessary-but-INSUFFICIENT. With the cap bumped to 8192, the warmup turn (~10 positions, far under 8192) STILL crashes. The FIRST mid-capture failure in the journal is NOT the KV:
```
fak-cuda: cudaMalloc(20480 bytes) failed: operation not permitted when stream is capturing
inkernel_chat oom-retry ... class=unknown site=dalloc bytes=20480 action=trim-idle-pools   <-- FIRST
... then ...
inkernel_chat oom-retry ... class=kv_cache site=kv-key-prealloc layer 0 bytes=33554432     <-- cascade
```
**20480 bytes = 5120 (hidden dim) × 4 = one decode-path activation/scratch buffer** (`site=dalloc class=unknown`), allocated fresh DURING the captured token because the buffer pool was not warm for it. That cudaMalloc-during-capture panics; the panic unwinds **past `finishGraph()`** (hal.go:314-322) so `GraphEndLaunch` (the `cudaStreamEndCapture`) never runs → **the capture is LEAKED (stream stays in capture mode)**. Every later alloc — including the KV prealloc — then fails 'operation not permitted when stream is capturing'. The visible `kv-key-prealloc` line is the cascade, not the trigger.

The HAL's own precondition (hal.go:367) is 'pool warm (no cudaMalloc during capture)'; the `s.halStep >= 2` gate (hal.go:306) does NOT satisfy it — 2 warmup steps don't allocate the decode activation buffers, so the first captured decode token mallocs.

**The complete fix is three parts:**
1. **Fully warm the buffer pool before the FIRST GraphBegin** — run >=1 full non-graph DECODE step (not just prefill) so every per-token activation/scratch size is resident, OR pre-touch them. (The current halStep>=2 gate is the bug.)
2. **Leak-defer hardening** — a `defer` on the captured-token path that, on panic, aborts the capture (cudaStreamEndCapture + discard) + GraphReset so one mid-capture malloc can't leak the stream and poison every later request. (Self-healing; turns a hard crash into a one-token non-graph fallback.)
3. **Size cudaKVMaxPos to --context-budget-tokens** (the plumbing patch already drafted) so AppendKV never reallocs mid-capture for long contexts.

Witnessed: 14B baseline 6.7 tok/s; with all three, graph replay (#483) is modeled ~8-15x. The cap-only change was built + tested on the A100 and reproduced the crash, isolating (1) as the real trigger.

**Comment by anthony-chaudhary** [comment]
Threaded under the GPU decode-parity playbook epic **#977**.

This is the CUDA instance of the playbook's **step 1** ("one submit per forward"): the CUDA graph is exactly the one-command-buffer lever that #67 paid on Metal via `MTLCommandBuffer`. The 27B-on-A100 witness here is what turns the gated-off `graphEnabled` default into a measured before/after — discard the cold token, `CUDA_GRAPH=0` vs `=1`, record under `experiments/gpu/`, each step naming the diagnostic. Blocked first by #969 (the capture crash); decode must also be byte-matched (#949 raw Q4_K) so the number is device-matched. Once graph-on is witnessed, #977's "find the serial step" lever applies to whatever stage still leaves the memory system idle.

**Comment by anthony-chaudhary** [comment]
Resolved by `d7d6172` (resolving commit). Closed by the DOS dispatch loop's close-resolved arm, witnessed via `dos commit-audit` (verdict OK / diff-witnessed). Reopen if this does not fully resolve it.

**Comment by anthony-chaudhary** [comment]
Fixed and witnessed on a live A100-40GB (native CUDA, 2026-06-27), commit b3d19d36 (branch wip932-graphfix).

Root cause confirmed live: the first **captured** logits step did a `cudaMalloc` mid-capture — it uploaded `model.norm` + `lm_head` and allocated the vocab-wide logits buffer, none of which the prefill `noLogits` steps warmed (they return before the logits path), and the C pool keys by exact byte size and falls through to `cudaMalloc` on a miss → illegal during capture → leaked the open capture → cascade.

Fix (3 parts): (1) gate capture on `halLogitsWarm` — capture only a logits step, and only after one ran **uncaptured** to pool `model.norm`/`lm_head` + the vocab logits transient; (2) `GraphAbort`/`fcuda_graph_abort` end+discard an open capture on the panic path so a stray mid-capture alloc can't leave the shared stream capturing; (3) `SetCUDAGraphKVCapacity` sizes the KV prealloc to `--context-budget-tokens`.

Witness: `fak serve --backend cuda --gguf <14B Q4_K> --cuda-graph` now serves correct code with **no crash** across 3 runs (previously crashed on first decode). No `ExecUpdate failed`/re-instantiate lines in the log → the graph is captured and replayed cleanly every token.
